actionpack/lib/action_dispatch/middleware/server_timing.rb

```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
+
+module ActionDispatch
+  class ServerTiming
+    class << self
+      def add_timing(key:, dur: nil, desc: nil)
+        timings[key] = { dur: dur, desc: desc }
+      end
+
+      def to_header
+        headers = []
+        timings.each_pair do |key, data|
+          header = "#{key}"
+          header += ";desc=\"#{data[:desc]}\"" if data[:desc]
+          header += ";dur=#{data[:dur]}" if data[:dur]
+          headers << header
+        end
+        headers.join(", ")
+      end
+
+      def measure(key:, desc: nil, &blk)
+        ActiveSupport::Notifications.instrument("measure.server_timing", desc: desc, key: key) do
+          yield
+        end
+      end
+
+      def clear_timings!
+        Thread.current[:server_timing_registry] = {}
+      end
+
+      def timings
+        Thread.current[:server_timing_registry] ||= {}
+      end
+    end
+
+    def initialize(app, default_events = [])
+      @app = app
+      subscribe_to_events(default_events) unless default_events.empty?
+    end
+
+    def call(env)
+      @app.call(env).tap do |response|
+        add_timing_headers(response[1])
+      end
+    end
+
+    private
+
+      def add_timing_headers(headers)
+        headers["Server-Timing"] = self.class.to_header
+      end
+
+      def subscribe_to_events(events)
```

Comment:
It'd be good if the user had a way to tell what kind of events he wants to subscribe to as well as what timing for each events.


Maybe the events could be formatted like this

```ruby
events = {
  'process_action.action_controller' => [:db_runtime, :view_runtime]
}

events.each do |namespace, events|
  AS.subscribe ...
end
```

Reply:
I'm not so keen on this idea because on this [list](https://guides.rubyonrails.org/active_support_instrumentation.html) I only see two timing metrics that we expose from rails: `db_runtime`, `view_runtime` and those both come from `process_action.action_controller`.   So I'm not sure what the extra flexibility gives us since there are no other particularly interesting metrics atm. 

If the user wants to subscribe to a custom notification they can do: 

``` ruby
ActiveSupport::Notifications.subscribe("my_action.my_namespace") do |*args|
  ServerTiming.add_timing(key: :some_key, description: :some_description)
end
```

This method is more intended to allow the user to choose wether they want to include these two `action_controller` events in their response.